### ISSUE TITLE
ShadyDOM Cannot read property '__catchall' of null

### ISSUE THREAD

**Issue by azakus** [webcomponents/shadydom#168]
_From @gaarf on July 19, 2017 16:30_

With `window.ShadyDOM = {force: true};` and webcomponentsjs@1.0.3, seeing the following:

```
attach-shadow.js:186 Uncaught TypeError: Cannot read property '__catchall' of null
    at DocumentFragment.l.f (attach-shadow.js:186)
    at DocumentFragment.l.v (attach-shadow.js:130)
    at DocumentFragment.l._renderRoot (attach-shadow.js:117)
    at DocumentFragment.l.Ea (attach-shadow.js:110)
    at attach-shadow.js:76
    at na (flush.js:28)
    at HTMLElement.dispatchEvent (patch-builtins.js:97)
    at dispatchNotifyEvent (property-effects.html:276)
    at Object.runNotifyEffect [as fn] (property-effects.html:300)
    at runEffectsForProperty (property-effects.html:137)
```

_Copied from original issue: webcomponents/webcomponentsjs#813_

**Comment by azakus** [comment]
_From @gaarf on July 20, 2017 17:45_

possibly related to https://github.com/webcomponents/shadydom/issues/166

**Comment by azakus** [comment]
@gaarf Can you provide a reproduction that triggers this?

**Comment by gaarf** [comment]
@azakus haven't been able to reduce yet, but noticed if I comment out `paper-tabs`, I only get one error thrown instead of dozens. Will keep trying to make a codepen.

**Comment by gaarf** [comment]
maybe it's a bug with `paper-tabs` ? here's a reproduction: http://jsbin.com/yikeyojilu/1/edit?html,console,output

**Comment by JoelCode** [comment]
chrome 60 
Cannot read property '__catchall' of null

chrome 60                Tests failed: 1 failed tests
firefox 54               Tests passed
IE 11.0.14393.953        Tests passed
Test run ended in failure: 1 failed tests

chrome 60 (2/0/1) 
firefox 54 (2/0/0)
IE 11.0.14393.953 (2/0/0)

```
      test() {
        return 'yes';
      }
```
I tried assert, expect and should but it's all fail.

**Comment by awikkerink** [comment]
I can reproduce with an element that has the `iron-resize-behavior` or has properties with `notify: true`.

**Comment by awikkerink** [comment]
This triggers the error:

Polymer#2.0.2
webcomponentsjs#1.0.5

```html
<link rel="import" href="../polymer/polymer.html">

<dom-module id="reproduce-inner">
        <template></template>
        <script>
                Polymer({
                        is: 'reproduce-inner',
                        properties: {
                                test: {
                                        type: String,
                                        notify: true,
                                        value: 'test'
                                }
                        }
                });
        </script>
</dom-module>

<dom-module id="reproduce-component">
        <template>
                <reproduce-inner>
                        <slot></slot>
                </reproduce-inner>
        </template>
        <script>
                Polymer({
                        is: 'reproduce-component'
                });
        </script>
</dom-module>
```

```html
<html>
	<head>
		<script src="../webcomponentsjs/webcomponents-lite.js"></script>
		<link rel="import" href="component.html">
	</head>
	<body>
		<reproduce-component><div></div></reproduce-component>
	</body>
</html>
```

**Comment by halkeye** [comment]
I don't think its chrome 60 related.

https://saucelabs.com/beta/tests/ecb5107625644fb4ad4f66ba581ff186 - Firefox 54 - pass
https://saucelabs.com/beta/tests/219e4a7accc14b3a968225e4e09c6151 - Chrome 58 - fail
https://saucelabs.com/beta/tests/e13d4eeddf5b4a039b046ce59a4319ce - Chrome 59 - fail
https://saucelabs.com/beta/tests/9e8e062b668f47f18879e97486dfa0ce - Chrome 60 - fail

Its not the exact same test, but it was having the same error message.

**Comment by bashmish** [comment]
Also have this issue. Do not have an easy example yet, but looks like it is related to events firing before the slot content is initialized. Because I clearly see this issue if I have `notify: true` on some properties and/or `this.fire()` calls in the element that has a slot wrapped within `dom-if`.

**Comment by juicelink** [comment]
I've 2 issues related to this one in IE11, related to slot utilisation
- first: _slotMap is initialised to null, so _slotMap['__catchall'] throw an exception.
- second: when creating a new Event('slotchanged'), IE11 complain that the constructor method is'nt supported

**Comment by srijken** [comment]
Confirmed that rollback to v1.0.0 fixes this problem.

If you inspect this code, it seems that if _slotMap would be defined, there would still be a problem assigning a slot: https://github.com/webcomponents/shadydom/blob/v1.0.1/src/attach-shadow.js#L188-L194

There's probably something wrong in the initialization then?

**Comment by bennypowers** [comment]
In the mean time

```js
// test/index.html 
    const suites = [
        'eve-upload-url_test.html?dom=shadow',
        'eve-upload-url_test.html?wc-shadydom=true',
        'eve-upload_test.html?dom=shadow',
        'eve-upload_test.html?wc-shadydom=true',
      ];

      const CAN_SHADOW_DOM = (document.head.createShadowRoot || document.head.attachShadow); // eslint-disable-line max-len
      suites = CAN_SHADOW_DOM ? suites : suites.filter((t) => t.includes('wc-shadydom'));
      function isChrome() {
        let isChromium = window.chrome;
        let winNav = window.navigator;
        let vendorName = winNav.vendor;
        let isOpera = winNav.userAgent.indexOf('OPR') > -1;
        let isIEedge = winNav.userAgent.indexOf('Edge') > -1;
        let isIOSChrome = winNav.userAgent.match('CriOS');

        if (isIOSChrome) {
          return true;
        } else if (
          isChromium !== null &&
          typeof isChromium !== 'undefined' &&
          vendorName === 'Google Inc.' &&
          isOpera === false &&
          isIEedge === false
        ) {
          return true;
        } else {
          return false;
        }
      }

      suites = isChrome() ? suites.filter((t) => !t.includes('shady')) : suites;
      WCT.loadSuites(suites);

```

**Comment by BorntraegerMarc** [comment]
I'm having the same issue with Edge 14 + 15

**Comment by BorntraegerMarc** [comment]
Can we expect a solution to this soon? It's blocking our upgrade of polymer as we now have to use webcomponentsjs 1.0.1

**Comment by azakus** [comment]
Actually, this looks a bit more complicated than what I fixed in #181. It looks like any event being fired during a custom element's `connectedCallback` can break the slot distribution algorithm.

Thankfully, this is pretty easy to reproduce with @awikkerink's example, and this is top priority!
